From 882efd8d555e5705f080e339f8f3c4c89cfcf4b3 Mon Sep 17 00:00:00 2001
From: Eberhard Beilharz <eb1@sil.org>
Date: Wed, 14 Dec 2011 09:37:00 +0100
Subject: [PATCH 2/3] bug-xamarin-2392: Adjust row height in DataGridView

A DataGridView in VirtualMode has to call OnRowHeightInfoNeeded and
update the height of the rows when adding new rows.

Also added a unit test that demonstrates the problem.

Change-Id: I4c28ac44fbdeb4c311aeded62e845d89f3c03551
---
 .../System.Windows.Forms/DataGridView.cs           |   15 +++++++++++++--
 .../Test/System.Windows.Forms/DataGridViewTest.cs  |   17 +++++++++++++++++
 2 files changed, 30 insertions(+), 2 deletions(-)

Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridView.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridView.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridView.cs
@@ -1036,8 +1036,14 @@
 				} else if (value > rows.Count) {
 					// If we need to add rows and don't have any columns,
 					// we create one column
-					if (ColumnCount == 0)
-						ColumnCount = 1;
+					if (ColumnCount == 0) {
+						System.Diagnostics.Debug.Assert (rows.Count == 0);
+						ColumnCount = 1; // this creates the edit row
+						if (VirtualMode) {
+							// update edit row height
+							UpdateRowHeightInfo (0, false);
+						}
+					}
 
 					List<DataGridViewRow> newRows = new List<DataGridViewRow> (value - rows.Count);
 					for (int i = rows.Count; i < value; i++)
@@ -5072,6 +5078,11 @@
 			if (hover_cell != null && hover_cell.RowIndex >= e.RowIndex)
 				hover_cell = null;
 
+			if (VirtualMode) {
+				for (int i = 0; i < e.RowCount; i++)
+					UpdateRowHeightInfo (e.RowIndex + i, false);
+			}
+
 			// Select the first row if we are not databound. 
 			// If we are databound selection is managed by the data manager.
 			if (IsHandleCreated && DataManager == null && CurrentCell == null && Rows.Count > 0 && Columns.Count > 0)
Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/Test/System.Windows.Forms/DataGridViewTest.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/Test/System.Windows.Forms/DataGridViewTest.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/Test/System.Windows.Forms/DataGridViewTest.cs
@@ -1879,6 +1879,23 @@
 				Assert.AreEqual (30, dgv.Rows [0].MinimumHeight);
 			}
 		}
+
+		[Test]
+		public void RowHeightInVirtualMode ()
+		{
+			using (var dgv = new DataGridView ()) {
+				dgv.RowHeightInfoNeeded += (sender, e) => {
+					e.Height = 50;
+					e.MinimumHeight = 30;
+				};
+				dgv.VirtualMode = true;
+				dgv.RowCount = 2;
+				Assert.AreEqual (50, dgv.Rows [0].Height);
+				Assert.AreEqual (30, dgv.Rows [0].MinimumHeight);
+				Assert.AreEqual (50, dgv.Rows [1].Height);
+				Assert.AreEqual (30, dgv.Rows [1].MinimumHeight);
+			}
+		}
 	}
 	
 	[TestFixture]
